Add ZHA configuration number entity to set/get thermostat cluster local temperature offset

theorlangur opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
I don't expect anything to break. 1 new configuration number entity for a ZHA device implementing a thermostat cluster may be added.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

I have a Sonoff TRVZB radiator valve regulator. This device exposes a thermostat cluster and among other things allows offsetting the temperature from the local sensor according to ZCL (at least) in a range -2.5 - 2.5 degree celsius.
It would be cool to have access to this configuration attribute via a separate entity in HA (and not only with a low-level write/read zigbee attribute calls). I tried it on my system and it seems to work and it makes the device actually change the reported values for a local temperature sensor.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
per ZCL 6.3.2.2.2 Thermostat Settings Attribute Set attribute id 0x0010 'LocalTemperatureCalibration', range -2.5 - 2.5 degree celsius

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

theorlangur committed:
adds a thermostat configuration entity to set/get local_temperature_calibration attribute

per ZCL 6.3.2.2.2 Thermostat Settings Attribute Set
attribute id 0x0010 'LocalTemperatureCalibration', range -2.5 - 2.5
degree celsius

elupus reviewed (dismissed):
You have unrelated changes in this pull

theorlangur commented:
> You have unrelated changes in this pull

@elupus , yes, sorry. I'm not sure how it got there. This was definitely not intended

dmulcahey commented on homeassistant/components/zha/number.py:
```diff
@@ -37,6 +38,7 @@
 CONFIG_DIAGNOSTIC_MATCH = functools.partial(
     ZHA_ENTITIES.config_diagnostic_match, Platform.NUMBER
 )
+MULTI_MATCH = functools.partial(ZHA_ENTITIES.multipass_match, Platform.NUMBER)
```
not needed

theorlangur replied:
addressed with
https://github.com/home-assistant/core/pull/105765/commits/3435b63b82ce963cfc9c606cb9aadf91fe04a695

dmulcahey commented on homeassistant/components/zha/number.py:
```diff
@@ -947,3 +949,25 @@ class AqaraThermostatAwayTemp(ZHANumberConfigurationEntity):
     _attr_icon: str = ICONS[0]
+
+
+@MULTI_MATCH(
```
This should use CONFIG_DIAGNOSTIC_MATCH defined on line 38

theorlangur replied:
addressed with
https://github.com/home-assistant/core/pull/105765/commits/3435b63b82ce963cfc9c606cb9aadf91fe04a695

dmulcahey reviewed (changes requested):
remove multi match and use CONFIG_DIAGNOSTIC_MATCH as this is a config entity

dmulcahey commented on homeassistant/components/zha/number.py:
```diff
@@ -947,3 +949,25 @@ class AqaraThermostatAwayTemp(ZHANumberConfigurationEntity):
+
+@MULTI_MATCH(
+    cluster_handler_names=CLUSTER_HANDLER_THERMOSTAT,
+    stop_on_match_group=CLUSTER_HANDLER_THERMOSTAT,
```
stop_on_match_group is meant to prevent different implementations for the same functionality from being matched. Do you need this for this entity?

theorlangur replied:
addressed with
https://github.com/home-assistant/core/pull/105765/commits/3435b63b82ce963cfc9c606cb9aadf91fe04a695

theorlangur committed:
removed MULTI_MATCH, using CONFIG_DIAGNOSTIC_MATCH

theorlangur commented:
@dmulcahey , thanks for your time and your input!
I'm obviously new to HA development and was looking at some examples to get what I wanted working. Apparently I misunderstood the meaning of MULTI_MATCH and what was actually needed in my case.
Thanks for pointing that out, I've removed MULTI_MATCH and using CONFIG_DIAGNOSTIC_MATCH as suggested. 
It's working locally for me.
I have a question though: how can I provide changes for translations/en.json? If I should at all.

dmulcahey commented:
> I have a question though: how can I provide changes for translations/en.json? If I should at all.

what do you mean?

theorlangur commented:
> > I have a question though: how can I provide changes for translations/en.json? If I should at all.
> 
> what do you mean?

I mean I've changed 'zha/number.py', I added a new entry to 'zha/strings.json'. As I understand most (all?) of the strings in strings.json are also present in 'zha/translations/en.json', so I thought about adding an entry there too, but realized that all translations are in .gitignore.
So my question was: do I still need to add it somehow (via some other mechanism) there? Or does it happen "automagically" via other means?
But I guess I'm proceeding with this PR in the meantime

puddly commented:
`strings.json` is all you need. `en.json` can be generated for local testing with the following command:

```
python3 -m script.translations develop
```

The other translations will be downloaded automatically and included in Home Assistant releases, they're not maintained in-tree but contributed through [Lokalise](https://developers.home-assistant.io/docs/translations/).

theorlangur commented on homeassistant/components/zha/number.py:
```diff
@@ -947,3 +948,22 @@ class AqaraThermostatAwayTemp(ZHANumberConfigurationEntity):
     _attr_icon: str = ICONS[0]
+
+
+@CONFIG_DIAGNOSTIC_MATCH(cluster_handler_names=CLUSTER_HANDLER_THERMOSTAT)
```
But this is a standard Zigbee attribute as described in ZCL 6.3.2.2.2.1 (Thermostat Settings Attribute Set attribute id 0x0010 'LocalTemperatureCalibration')
What would be specified as a model in the matcher?

dmulcahey replied:
all good I just hadn't checked the spec

dmulcahey commented on homeassistant/components/zha/number.py:
```diff
@@ -947,3 +948,22 @@ class AqaraThermostatAwayTemp(ZHANumberConfigurationEntity):
+    """Local temperature calibration."""
+
+    _unique_id_suffix = "local_temperature_calibration"
+    _attr_entity_category = EntityCategory.CONFIG
```
this is inherited. No need to redefine it

theorlangur replied:
"_attr_entity_category", right?

dmulcahey replied:
yes

theorlangur replied:
no problem
https://github.com/home-assistant/core/pull/105765/commits/b8b2e14443c47d86abb2d79c292a330d0233a4f1

dmulcahey commented:
one last comment that I missed before then this is good to go. Thanks for the contribution!!

theorlangur committed:
removed explicit '_attr_entity_category', since it's inherited

dmulcahey commented:
ugh missed one more thing... sorry. You need to add local temperature calibration to `ZCL_INIT_ATTRS` here: https://github.com/home-assistant/core/blob/b8b2e14443c47d86abb2d79c292a330d0233a4f1/homeassistant/components/zha/core/cluster_handlers/hvac.py#L103

sorry about that!

theorlangur committed:
added "local_temperature_calibration" attribute to ZCL_INIT_ATTRS

dmulcahey commented:
> > ugh missed one more thing... sorry. You need to add local temperature calibration to `ZCL_INIT_ATTRS` here:
> > https://github.com/home-assistant/core/blob/b8b2e14443c47d86abb2d79c292a330d0233a4f1/homeassistant/components/zha/core/cluster_handlers/hvac.py#L103
> > 
> > sorry about that!
> 
> Nothing to be sorry about) What does 'bool' mean there? Like if an attribute is required for a cluster or what? In this case it should be 'False', I guess?

It should be True.

theorlangur commented:
> It should be True.

should I also add a property getter for it (like as I can see some exist for other attributes)?

theorlangur commented:
> It should be True.

https://github.com/home-assistant/core/pull/105765/commits/964265271292a363768477f9b863674571712b8f

dmulcahey commented:
> > It should be True.
> 
> should I also add a property getter for it (like as I can see some exist for other attributes)?

no need

dmulcahey commented:
> > ugh missed one more thing... sorry. You need to add local temperature calibration to `ZCL_INIT_ATTRS` here:
> > https://github.com/home-assistant/core/blob/b8b2e14443c47d86abb2d79c292a330d0233a4f1/homeassistant/components/zha/core/cluster_handlers/hvac.py#L103
> > 
> > sorry about that!
> 
> Nothing to be sorry about) What does 'bool' mean there? Like if an attribute is required for a cluster or what? In this case it should be 'False', I guess?

and to close the loop on the question: https://github.com/home-assistant/core/blob/093c952c38fb25e9046b36d219c182479c376ccd/homeassistant/components/zha/core/cluster_handlers/__init__.py#L130

it's says that we can rely on cache instead of trying to live read the device on every startup

theorlangur commented:
> https://github.com/home-assistant/core/blob/093c952c38fb25e9046b36d219c182479c376ccd/homeassistant/components/zha/core/cluster_handlers/__init__.py#L130
> 
> it's says that we can rely on cache instead of trying to live read the device on every startup

Got it
Thanks for clarifications!

dmulcahey commented:
@elupus i can’t seem to dismiss your review. This should be good to merge now.
